### ISSUE TITLE
Wrongly mark a provider usage as deprecated when used via the dependency engine

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#15146]
### Expected Behavior

Gradle should not nag a deprecation warning for a provider that is "query" via the dependency engine. In fact, in the sample shown, Gradle should only be visiting the provider to identify the implicit task dependencies and not access the value.

### Current Behavior

Gradle somehow triggers a provider to be realized when building the task graph.

### Context

This issue affects the usability of the Provider API in the dependency engine. We can work around the issue by using `forUseAtConfigurationTime()` but this workaround is completely wrong. We aren't using the provider at configuration time, it's Gradle that is either over-using the provider or wrongly attributing its own usage of the provider for building the task graph as deprecated.

### Steps to Reproduce 

Use the following code with two projects:
```
project(':consumer') {
    configurations {
        foo {
            canBeResolved = true
            canBeConsumed = false

            attributes {
                attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, 'foo'))
            }
        }
    }

    dependencies {
        foo project(':producer')
    }

    tasks.register('foo') {
        inputs.files(configurations.foo)
    }
}

project(':producer') {
    def generateSources = tasks.create('generateSources') {
        def outputDirectory = layout.buildDirectory.dir('generated-srcs').get().asFile
       outputs.dir(outputDirectory) 
       doLast {
            new File(outputDirectory, 'foo') << 'bar'
       }
    }

    def sources = files(generateSources)

    configurations {
        elements {
            canBeResolved = false
            canBeConsumed = true

            attributes {
                attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, 'foo'))
            }

            def directoryProvider = sources.elements.map { it.first().asFile }
            outgoing.artifact(directoryProvider)
        }
    }
}
```

### Your Environment

Build scan URL: https://gradle.com/s/pn3ketnudqusg

**Comment by lacasseio** [comment]
The only workaround found may actually expose another bug or shortcoming of the Provider API and what the deprecation warning is trying to prevent. We can "erase" the task dependency from the provider by flat mapping the `sources.elements` in the producer's outgoing artifact and attach the original provider via `builtBy`:

```
            def directoryProvider = sources.elements.flatMap { list -> providers.provider {list.first().asFile} }
            outgoing.artifact(directoryProvider) {
                builtBy sources
            }
```

Although this works, it actually jumps over the restriction that Gradle tries to impose that is not querying the output value of a task before it runs. I'm unsure if this would be the right intention.

**Comment by lacasseio** [comment]
But if the provider gets mapped before hand then the work around does not work:

```
            def mappedProvider = sources.elements.map { it.first() }.map { it.asFile }
            def directoryProvider = mappedProvider.flatMap { file -> providers.provider {file} }
            outgoing.artifact(directoryProvider) {
                builtBy mappedProvider
            }
```

**Comment by lacasseio** [comment]
Is there any chance this will be addressed before 7.0?

**Comment by lacasseio** [comment]
Any chance this will ever be addressed? What would be the alternative if this is not an issue to be addressed?

**Comment by eskatos** [comment]
Hey Daniel,
I'm not sure but https://github.com/gradle/gradle/pull/15996 might have helped with this issue.
Could you please give `7.0-milestone-1` a try?

**Comment by lacasseio** [comment]
Maybe, I will need a bit of time to try it out I'm in the middle of some refactoring work.

**Comment by wolfs** [comment]
Are you looking for something like #20155 here? Or is this something different?

**Comment by lacasseio** [comment]
No, this was in part a demonstration that `FileCollection#getElements()` doesn't behave the same as `FileCollection`. In terms of outgoing `Configuration` we can't use a `FileCollection` hence we need individual files/directories to export. What we do at Nokee is use the elements provider to map the files into a lazy list of `PublishArtifact` which we add to the artifact domain object set (via `addAllLater`). The provided sample abstracts all this into a minimum reproducer. At the core, the provider passed to the `artifact` method has the dependency information and Gradle will rightly execute the dependency before, however, internally, the dependency engine will resolve the provider "early" causing the warning (and now crash on 7.x) which is wrong as everything will be done correctly in the end.

**Comment by eskatos** [comment]
Confirmed this is still the case with Gradle 8.2.1 but now it is a failure instead of a deprecation warning.

Complete reproducer available at https://github.com/eskatos/issue-reproducers/tree/main/gradle/issue-15146

https://scans.gradle.com/s/ycntgk44o5que/failure#1

**Comment by eskatos** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by mlopatkin** [comment]
@gradle/bt-jvm, sorry, this was stuck between the boards. Could you please take it from here and triage the issue?

**Comment by big-guy** [comment]
This is related https://github.com/gradle/gradle/issues/27578.  This describes some workarounds.

We need to decide how we would like this to work so that published artifacts carry the proper dependencies. 

We don't plan to work on this soon.
